File: notifications-api-dotnet/Processor.cs
using NotificationApi.Providers;

namespace NotificationApi;

// Define the interface for the different notification providers (Strategy Pattern)
public interface INotificationProvider
{
    ProviderResponse Send(string recipient, string message);
}

// Implement wrapper classes (Adapters) for external providers without modifying the original files
public class EmailNotificationProvider : INotificationProvider
{
    public ProviderResponse Send(string recipient, string message)
    {
        return EmailProvider.Send(new Dictionary<string, string>
        {
            { "recipient", recipient },
            { "message", message }
        });
    }
}

public class SmsNotificationProvider : INotificationProvider
{
    public ProviderResponse Send(string recipient, string message)
    {
        return SmsProvider.Send(new Dictionary<string, string>
        {
            { "recipient", recipient },
            { "message", message }
        });
    }
}

public class PushNotificationProvider : INotificationProvider
{
    public ProviderResponse Send(string recipient, string message)
    {
        return PushProvider.Send(new Dictionary<string, string>
        {
            { "recipient", recipient },
            { "message", message }
        });
    }
}

// The refactored and corrected NotificationProcessor class
public class NotificationProcessor
{
    private readonly Dictionary<string, INotificationProvider> _providers;

    public NotificationProcessor()
    {
        // Register providers flexibly using a dictionary to satisfy the Open/Closed Principle
        _providers = new Dictionary<string, INotificationProvider>(StringComparer.OrdinalIgnoreCase)
        {
            { "email", new EmailNotificationProvider() },
            { "sms", new SmsNotificationProvider() },
            { "push", new PushNotificationProvider() }
        };
    }

    public void SendOne(Notification n)
    {
        n.Status = NotificationStatuses.Processing;
        n.Attempts += 1;
        n.LastAttemptAt = DateTime.Now.ToString("O");

        if (n.TargetChannels == null || n.TargetChannels.Count == 0)
        {
            n.Status = NotificationStatuses.Failed;
            n.LastError = "No target channels";
            return;
        }

        bool allSentSuccessfully = true;
        List<string> errorMessages = new();

        // Fix 1: Loop through ALL requested channels instead of just the first one
        foreach (var target in n.TargetChannels)
        {
            if (!_providers.TryGetValue(target.Type, out var provider))
            {
                allSentSuccessfully = false;
                errorMessages.Add($"[{target.Type}] Unknown channel type");
                continue;
            }

            try
            {
                var response = provider.Send(target.Value, n.Message);

                // Fix 2: Evaluate the actual provider response instead of blindly marking as Sent
                if (response.Result == "error" || !string.IsNullOrEmpty(response.ErrorCode))
                {
                    allSentSuccessfully = false;
                    errorMessages.Add($"[{target.Type}] Error: {response.Message} (Code: {response.ErrorCode})");

                    // Handle temporary issues to allow subsequent retry attempts
                    if (response.ErrorCode == "temporary_outage" || response.Message.Contains("temporary", StringComparison.OrdinalIgnoreCase))
                    {
                        n.Status = NotificationStatuses.RetryPending;
                    }
                }
            }
            catch (Exception ex)
            {
                allSentSuccessfully = false;
                errorMessages.Add($"[{target.Type}] System Exception: {ex.Message}");
            }
        }

        // Update the final notification status based on the results of all targeted channels
        if (allSentSuccessfully)
        {
            n.Status = NotificationStatuses.Sent;
            n.LastError = "All channels delivered successfully";
        }
        else
        {
            // If the status wasn't explicitly set to RetryPending during the loop, mark it as Failed
            if (n.Status != NotificationStatuses.RetryPending)
            {
                n.Status = NotificationStatuses.Failed;
            }
            n.LastError = string.Join(" | ", errorMessages);
        }
    }

    public void SendAll()
    {
        // Process both Pending notifications and those waiting for a retry execution
        var pending = Storage.Notifications
            .Where(n => n.Status == NotificationStatuses.Pending || n.Status == NotificationStatuses.RetryPending)
            .ToList();

        foreach (var n in pending)
        {
            SendOne(n);
        }
    }
}

File: src/processor.ts
import {
  Notification,
  PENDING,
  PROCESSING,
  SENT,
  FAILED,
  RETRY_PENDING,
} from "./models.js";
import * as storage from "./storage.js";
import { send as sendEmail } from "./providers/emailProvider.js";
import { send as sendSms } from "./providers/smsProvider.js";
import { send as sendPush } from "./providers/pushProvider.js";

export interface ProviderRequest {
  recipient: string;
  message: string;
}

export interface ProviderResponse {
  Result: string;
  ErrorCode: string;
  Message: string;
}

// A provider is just a function from a request to a response. Modelling it as a
// registry (rather than a hardcoded if/else) keeps the processor open for new
// channels and lets tests inject deterministic fakes.
export type Provider = (req: ProviderRequest) => ProviderResponse;

export const defaultProviders: Record<string, Provider> = {
  email: sendEmail,
  sms: sendSms,
  push: sendPush,
};

type Outcome = "ok" | "temporary" | "permanent";

// Map a provider's Result to a delivery outcome. Anything that isn't an explicit
// success or a known-temporary failure is treated as permanent (fail closed).
function classify(result: string): Outcome {
  if (result === "Success") return "ok";
  if (result === "TemporaryFailure") return "temporary";
  return "permanent";
}

export class NotificationProcessor {
  constructor(private readonly providers: Record<string, Provider> = defaultProviders) {}

  sendOne(n: Notification): void {
    n.status = PROCESSING;
    n.attempts++;
    n.lastAttemptAt = new Date();

    if (n.targetChannels.length === 0) {
      n.status = FAILED;
      n.lastError = "No target channels";
      return;
    }

    let anyPermanent = false;
    const errors: string[] = [];

    // Attempt every channel, not just the first, and collect outcomes.
    for (const channel of n.targetChannels) {
      const provider = this.providers[channel.type];
      if (!provider) {
        anyPermanent = true;
        errors.push(`[${channel.type}] unknown channel type`);
        continue;
      }

      const response = provider({ recipient: channel.value, message: n.message });
      const outcome = classify(response.Result);
      if (outcome === "ok") continue;

      if (outcome === "permanent") anyPermanent = true;
      errors.push(`[${channel.type}] ${response.Message}`.trim());
    }

    if (errors.length === 0) {
      n.status = SENT;
      n.lastError = null;
    } else {
      // Permanent failures dominate: retrying can't fix a hard bounce, so the
      // notification is failed rather than parked for retry. Any error with no
      // permanent failure is temporary-only, hence retry_pending. (Trade-off
      // recorded in NOTES.md.)
      n.status = anyPermanent ? FAILED : RETRY_PENDING;
      n.lastError = errors.join(" | ");
    }
  }

  sendAll(): void {
    // Drain both fresh and retry-eligible notifications.
    const due = storage
      .getAll()
      .filter((n) => n.status === PENDING || n.status === RETRY_PENDING);
    for (const n of due) {
      this.sendOne(n);
    }
  }
}
